traits/trait_list_object.py

```diff
@@ -53,460 +94,754 @@ def __init__(self, index=0, removed=None, added=None):
+
+
+        Notes
+        -----
+        Notification:
+            index : int
+                The index before which the item was inserted.
+            added : list
+                The newly added item(s) as a list.
+            removed : list
+                Will be []
+
+        """
+        removed = []
+        norm_index = self._normalize_index(index)
+        added = self.validate(norm_index, removed, [object])
 
-        if trait.minlen <= (len(self) + len_xlist) <= trait.maxlen:
-            object = self.object()
-            name = self.name
-            validate = trait.item_trait.handler.validate
-            try:
-                if validate is not None:
-                    xlist = [validate(object, name, value) for value in xlist]
+        super().insert(index, added[0])
 
-                list.extend(self, xlist)
+        if self._should_notify(removed, added):
+            self.notify(norm_index, removed, added)
+
+    def clear(self):
+        """ Remove all items from the list
 
-                if (self.name_items is not None) and (len(xlist) != 0):
-                    self._send_trait_items_event(
-                        self.name_items,
-                        TraitListEvent(len(self) - len(xlist), None, xlist),
-                        trait.items_event(),
-                    )
 
-                return
+        Notes
+        -----
+        Notification:
+            index : slice
+                The range of the slice will be 0 to len(self).
+            added : list
+                The empty list [].
+            removed : list
+                The removed items.
+
+        """
+        index = slice(0, len(self), None)
+        removed = self.copy()
+        self.validate(index, removed, [])
+        super().clear()
+
+        if self._should_notify(removed, []):
+            self.notify(index, removed, [])
+
+    def pop(self, index=-1):
+        """ Remove and return item at index (default last).
+
+        Parameters
+        ----------
+        index: int
+            Index at which item has to be removed.
+
+        Returns
+        -------
+        item : An item in the list at given index, last item if no index given.
+
+        Raises
+        ------
+        IndexError
+            If list is empty or index is out of range.
+
+        Notes
+        -----
+        Notification:
+            index : int
+                The normalized index between 0 and len(self).
+            added : list
+                Will be []
+            removed : list
+                The removed item, as a list.
+
+        """
 
-            except TraitError as excp:
-                excp.set_prefix("The elements of the")
-                raise excp
+        added = []
+        removed = self._get_removed(index, default=[])
+        norm_index = self._normalize_index(index)
+        self.validate(norm_index, removed, added)
 
-        self.len_error(len(self) + len(xlist))
+        removed_item = super().pop(index)
```

Comment:
It looks as though if the index is out of range, we still make a call to `self.validate` (with both the `added` and `removed` lists being empty). Is that the behaviour that we want? I think I'd expect no call to `self.validate` for an out-of-range index.

Reply:
My understanding is that this is all for getting the original error message from Python list (per our slack discussion). An alternative would be
```
try:
    removed = _get_removed(index) # Not leniently returning a default value
except BaseException:
    # to get the error message from Python
    super().pop(index)
    # we should not have got here... the unexpected modification is not undone however
   raise
else:
    self.validate(...)
    super().pop(index)
    self.notify(...)

```

Reply:
Right, we're juggling a number of constraints here. This would be much easier if all we had to do was validate individual incoming items. But I do think we shouldn't be calling `validate` on operations that already invalid at Python level.

Reply:
Is there any consensus on calling `validate` on invalid python operations ? I suspect some regression tests which check for the `TraitError` exception will have to be adjusted.

Reply:
This was discussed in other channels and the decision was to let this go for now. It's still desirable not to call the validation in cases where the base list operation would have raised, but it's proving awkward to make that unconditionally true.